Feature/NMD218-222 - Create exception middleware, Create "Services" folder, Create "Repositories" folder

Finka95 opened the pull request.

Finka95 committed:
start creating controllers for using EF

Commit:
create connection to database in controllers

Finka95 committed:
Add some business logic (e.g. if-statements, throw exception) in controllers.
Add another CRUD methods in controllers

Finka95 committed:
Add some business logic  in controllers

Finka95 committed:
add exception middlewaer

Finka95 committed:
Create "Services" folder and move business logic and database calling into this folder (start using DI)

Finka95 committed:
Create "Repositories" folder and move database calling to this folder (using DI)

Finka95 committed:
Add logger and change return type in methods of controllers

CrazyKamenwik commented on TicketSystem/TicketSystem/Data/Repositories/TicketRepository.cs:
```diff
@@ -0,0 +1,75 @@
+            return _context.Tickets.AsNoTracking().Where(conditions);
+        }
+
+        public async Task<int> CloseOpenTickets(int minutesToClose)
```
You have no int in the return clause. You can use just "Task" for such cases

v12code replied:
done

CrazyKamenwik commented on TicketSystem/TicketSystem/Data/Repositories/TicketRepository.cs:
```diff
@@ -0,0 +1,75 @@
+            return await _context.Tickets.AsNoTracking().ToListAsync();
+        }
+
+        public IQueryable<Ticket> GetUsersByConditionsAsync(Expression<Func<Ticket, bool>> conditions)
```
What kind of expression can be used as a parameter for this method?

Finka95 committed:
resolve conflicts

Finka95 committed:
Change the return type of CloseOpenTickets methods from Task<int> to Task

pavelGagarin0870 commented on TicketSystem/TicketSystem/Controllers/UserController.cs:
```diff
@@ -1,48 +1,49 @@
-            _logger.LogDebug("Post user");
-            await _context.Users.AddAsync(user);
-            await _context.SaveChangesAsync();
+            _logger.LogDebug("Put new user");
```
put to logs user object

pavelGagarin0870 commented on TicketSystem/TicketSystem/Controllers/UserController.cs:
```diff
@@ -70,7 +71,7 @@ public async Task<User> Post(User user)
             _context.Users.Remove(user);
             await _context.SaveChangesAsync();
 
```
Add logs for remove operations

pavelGagarin0870 commented on TicketSystem/TicketSystem/Data/Models/TicketStatus.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace TicketSystem.Data.Models
```
Move enums to separate folder (may name like "Enums")

pavelGagarin0870 commented on TicketSystem/TicketSystem/Data/Repositories/ITicketRepository.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace TicketSystem.Data.Repositories
+{
+    public interface ITicketRepository
```
Move abstractions to separate folders (create folder Abstractions, and inside sub folders Services, Repositories and etc.)

pavelGagarin0870 commented on TicketSystem/TicketSystem/Data/Models/Ticket.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+        }
+
+        public Ticket(ITicketService ticketService, User ticketCreator, User? freeOperator = null)
```
All logic should be considered in the services

Genresal commented on TicketSystem/TicketSystem/Controllers/MessageController.cs:
```diff
@@ -1,31 +1,30 @@
+            _messageService = messageService;
             _logger = logger;
             _context = new ApplicationContext();
         }
```
Is the _context variable still necessary there?

Genresal commented on TicketSystem/TicketSystem/Controllers/UserController.cs:
```diff
@@ -56,7 +57,7 @@ public async Task<User> Post(User user)
             _context.Entry(user).State = EntityState.Modified;
             await _context.SaveChangesAsync(); //DbUpdateConcurrencyException
 
```
The same question about _context.

Finka95 committed:
deleted _context from UserController

pavelGagarin0870 commented on TicketSystem/TicketSystem/Data/Repositories/IUserRepository.cs:
```diff
@@ -0,0 +1,14 @@
+    public interface IUserRepository
+    {
+        Task<User> CreateAsync(User user);
+        Task<User?> UpdateAsync(int id, User user);
```
You could put only model inside repo (inside controller add to Model id)

Finka95 committed:
- Change logs
- Move enum to separate folder
- Move abstractions to separate folders
- All logic should be considered in the services (Move timer from Ticket to TicketService)
- Change repositories methods for use onlu model in parameter

Finka95 committed:
Fixed bug with Ilogger in MessageController

pavelGagarin0870 commented on TicketSystem/TicketSystem/Data/Repositories/Abstractions/ITicketRepository.cs:
```diff
@@ -0,0 +1,15 @@
+        Task<Ticket?> UpdateAsync(Ticket ticket);
+        Task<Ticket?> DeleteAsync(int id);
+        Task<IEnumerable<Ticket>> GetAllAsync();
+        IQueryable<Ticket> GetUsersByConditionsAsync(Expression<Func<Ticket, bool>> conditions);
```
Why not iEnumerable

Add at the end of method .ToListAsync

Finka95 committed:
- Replased the LogDebug with LogInformation
- Add models as json in Logger in Post and Put methods
- Changed GetTicketsByConditionsAsync() method for return IEnumerable

pavelGagarin0870 commented on TicketSystem/TicketSystem/Controllers/MessageController.cs:
```diff
@@ -18,12 +20,11 @@ public MessageController(IMessageService messageService, ILogger<MessageControll
+        public async Task<Message> Post(int userId, Message message, CancellationToken cancellationToken)
         {
-            _logger.LogDebug("Post message bu userId {userId}", userId);
+            _logger.LogInformation("Post new message by userId {userId}, message: {JsonConvert.SerializeObject(message)}",userId, JsonConvert.SerializeObject(message));
```
message: {JsonConvert.SerializeObject(message)} instead of this could live just message: {message}

pavelGagarin0870 commented on TicketSystem/TicketSystem/Controllers/UserController.cs:
```diff
@@ -19,44 +21,43 @@ public UserController(IUserService userService, ILogger<UserController> logger)
+        public async Task<User> Post(User user, CancellationToken cancellationToken)
         {
-            _logger.LogDebug("Post new user");
+            _logger.LogInformation("Post new User {JsonConvert.SerializeObject(user)}", JsonConvert.SerializeObject(user));
```
same

pavelGagarin0870 commented on TicketSystem/TicketSystem/Controllers/UserController.cs:
```diff
@@ -19,44 +21,43 @@ public UserController(IUserService userService, ILogger<UserController> logger)
+        public async Task<User?> Put(int id, User user, CancellationToken cancellationToken)
         {
-            _logger.LogDebug("Put user by id {id}", id);
+            _logger.LogInformation("Put User with id {id}, user: {JsonConvert.SerializeObject(user)}",id, JsonConvert.SerializeObject(user));
```
same

pavelGagarin0870 commented on TicketSystem/TicketSystem/Data/Repositories/TicketRepository.cs:
```diff
@@ -15,62 +15,83 @@ public TicketRepository(ApplicationContext context)
-                .ToList();
+                                                          .First().User.UserRole == UserRole.Operator);
 
             foreach (var ticket in tickets)
```
Move business logic to services

pavelGagarin0870 commented on TicketSystem/TicketSystem/Data/Repositories/UserRepository.cs:
```diff
@@ -15,46 +16,74 @@ public UserRepository(ApplicationContext context)
-        public async Task<User?> UpdateAsync(User user)
+        public async Task<User?> UpdateAsync(User user, CancellationToken cancellationToken)
         {
             var userExist = _context.Users.Any(x => x.Id == user.Id);
```
Move this validation to bll

pavelGagarin0870 commented on TicketSystem/TicketSystem/Services/MessageService.cs:
```diff
@@ -15,14 +15,14 @@ public MessageService(IUserService userService, ITicketService ticketService)
 
             Ticket ticket;
-            var freeOperator = await _userService.GetFreeOperator();
+            var freeOperator = await _userService.GetFreeOperator(cancellationToken);
```
Free translated like "without cost", maybe better to rename this ?

Finka95 committed:
- Change GetFreeOperator to GetNotBusyOperator
- Change LogInformation in Post and Put methods
- Move userExist logic from repo to bll

Finka95 committed:
- Move userExist logic from dal to bll
